### ISSUE TITLE
workspace: symbolic link pointing to working directory

### ISSUE THREAD

**Issue by tiborsimko** [reanahub/reana-workflow-controller#235]
If the workflow need to access workspace directory as absolute path instead of relative path, the current technique isto use:

- `pushd .` at the beginning of step to store workspace location;
- do many things containing `cd` to special places and friends;
- `popd` to return back to workspace;
- `cp` any needed artefacts from special places into workspace.

For convenience, we could:

- create symbolic link `/workspace` for job pods pointing to the workspace directory,  so that users can always rely on this being the absolute path to the top directory of the workflow execution;

- set up WORKSPACE environment variable functioning the same.

- something else?

CC @clelange @lukasheinrich

**Comment by clelange** [comment]
CMSSW relies on an absolute path for its work area, and expects it not to change. If the work area is moved, it will complain and one will have to execute another command to update the path.
I can create a CMSSW work area e.g. in the form of `/workspace/CMSSW_X_Y`, but would this work with the above proposal?

**Comment by tiborsimko** [comment]
The "writable" area could be accessed as `/workspace`, and this would be symlinked to certain place (such as `/var/reana/users/<some-long-uuid>/workflows/4059f4da-0b2f-412b-9b5d-2ca16f5f0ff4`) for the whole duration of the workflow execution. All the jobs *within the same workflow* would share this workspace.
 
The "read-only" area where the code is hosted can be anywhere where the container wishes, e.g. `/code`, and this place will be also available to all the jobs *and also all the workflows* without any changes.

So, if CMSSW can be installed into a fixed place like `/code` or anywhere else like that, and if running the CMSSW software can be done elsewhere e.g. in the `/workspace` directory, then this would work well for the two scenarios, I think.

**Comment by tiborsimko** [comment]
BTW while we are at creating useful environment variables, we could create a bunch such as:

- `WORKFLOW_NAME` - the name of the currently running workflow e.g. `h4l`
- `WORKFLOW_RUN_NUMBER` - the run number e.g. `13`
- `WORKFLOW_WORKSPACE` - the full path to the workspace where workflow runs e.g. `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/4059f4da-0b2f-412b-9b5d-2ca16f5f0ff4`

This could be also useful for workflow execution to send alerts etc.

We could have later, if wanted, information about the user:

- `WORKFLOW_USER_UID`

or about currently running workflow step:

- `WORKFLOW_STEP_NAME`

etc. Only if wanted and if it makes sense.

Thinking about the other possible use cases will permit us to come up with nicely extensible naming scheme. 

P.S. We could use the same for CI e.g. create a bunch of `GITLAB_REPOSITORY`, `GITLAB_COMMIT`, `GITLAB_MERGE_REQUEST`, `GITLAB_BRANCH` etc to indicate which commit and/or which merge request triggered the CI build. CC @diegodelemos

**Comment by lukasheinrich** [comment]
I think indeed it would be useful to have a view on the data across all workflow processing steps that is

1. consistent across processing steps (on a absolute path level
2. independent from the rest of the innfrastructure (no workflow id's etc)

1. is important for cases where e.g. one step is preparing a list of files in a text file that a downstream step consumes
2. is important possibly for caching purposes, where you could cache across workflow runs

ideally I would not make the name too generic. It's easy to imagine some people preparinng their images in a way that they use "/data" or "/workdir" (in fact ATLAS does).. so maybe somethisng along the lines of "/reanadata" would work

**Comment by clelange** [comment]
> It's easy to imagine some people preparinng their images in a way that they use "/data" or "/workdir" (in fact ATLAS does).. so maybe somethisng along the lines of "/reanadata" would work

Knowing that ATLAS use `/data` and `/workdir`, we can also try to have this in CMS. Maybe we can come up with some guidelines? This should e.g. clarify what `/data` should be used for as opposed to `/reanadata`.
@lukasheinrich Can you list what you're using at the moment, and which of these are potentially writable directories or not?

**Comment by lukasheinrich** [comment]
so in ATLAS  / the grid container support (@afortiorama) we default to 

`/data` for the default location of input files etc
`/workdir` is the directory in which we see people putting their analysis code (it's also the WORKDIR of our base image) so that should *not* be overwritten

see a long discussion on  https://gitlab.cern.ch/atlas-sit/docker/merge_requests/35

we also default to a user called `atlas` inside of the images

**Comment by tiborsimko** [comment]
It is generally better to use relative paths rather than absolute path in order to ensure longer-term reproducibility of computational workflows. This is quite a well-established reproducibility principle that would be great to have more promotion for inside collaborations.  Even in the containerised world, see e.g. [CodeOcean capsule review lessons](https://medium.com/codeocean/five-reproducibility-lessons-from-a-year-of-reviewing-compute-capsules-de71729ebd8a). Moreover, in case of hybrid workflows, some computational steps might be dispatched to Condor or Slum, there might be no `/workdir` there to rely on, etc.

That said, we can come up with some recommendations on how to structure containers, the base image, the framework code, the user code, the input data and the output data -- perhaps as part of wider common "HEP Containers" effort?! Ideally addressing all these as a whole, so that if `/workdir` is being used for storing user code, then it won't be used to store temporarily produced data in the same directories, etc.

In any case, REANA won't override whatever convention the collaboration containers decide to use. It should offer an easy way to see which working directory the workflow started in. The easiest would be via `REANA_WORKFLOW_WORKSPACE` environment variable for example. Because the location could be Ceph shared filesystem location, or could be future S3 bucket or dCache completely outside of the container.

REANA is also rather neutral as to the container user -- provided the user belongs to the group that would have write permissions to the shared workspace.  This we are ensuring via *additional* GID=0 group membership similarly to what OpenShift does; see "Supporting arbitrary user IDs" in [Containerise your environment](https://reana.readthedocs.io/en/latest/userguide.html#containerise-your-environment) docs.

We can best muse about that lengthy ATLAS discussion and possible overlaps IRL :wink:

**Comment by afortiorama** [comment]
> It is generally better to use relative paths rather than absolute path in order to ensure longer-term reproducibility of computational workflows. This is quite a well-established reproducibility principle that would be great to have more promotion for inside collaborations.

that is true on bare metal, in containers it should be irrelevant because you port the container with the code in it. But even if you want to apply the same principle, as per the article you post, you need to decide where is the relative path, i.e. either in the directory where the code reside or in the directory where you do IO. In the lesson learned by running other people containers for testing it is much less error prone the latter.
